Added Dulmage Mendelsohn Decomposition

harmanpa opened the pull request:
Added the Dulmage Mendelsohn Decomposition to the connectivity algorithms.

This is a resubmission of the algorithm submitted in [#570](https://github.com/jgrapht/jgrapht/pull/570) to make a cleaner pull-request

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

jsichi commented:
In the review for #570, @jkinable suggested clarifying the class Javadoc as follows:

----

This class computes a <a href="https://en.wikipedia.org/wiki/Dulmage%E2%80%93Mendelsohn_decomposition">
 Dulmage-Mendelsohn Decomposition </a>` of a bipartite graph. A Dulmage–Mendelsohn decomposition is a
 partition of the vertices of a bipartite graph into subsets, with the property that two adjacent vertices belong to the
 same subset if and only if they are paired with each other in a <a 
href="http://mathworld.wolfram.com/PerfectMatching.html">perfect matching</a> of the graph. This particular 
implementation is capable of computing both a <i>coarse</i> and a <i>fine</i> Dulmage-Mendelsohn Decomposition.
 
<p>
A coarse Dulmage-Mendelsohn Decomposition is a partitioning... 
<p>
A fine Dulmage-Mendelsohn Decomposition is a partitioning...
<p>
The Dulmage-Mendelsohn Decomposition was introduced in:
Dulmage, A.L., Mendelsohn, N.S. Coverings of bipartitegraphs, Canadian J. Math., 10, 517-534, 1958.
The implementation of this class is based on:
Bunus P., Fritzson P., Methods for Structural Analysis and Debugging of Modelica Models, 2nd International Modelica Conference 2002<br>
The runtime complexity of this class is $O(V + E)$.

jsichi commented:
@jkinable had some other comments but I think you may have addressed them in this update.

d-michail commented:
There are multiple comments from #570 which still need addressing.

harmanpa commented:
I believe I had addressed the comments from #570 except for improving the description in the documentation, I've done this now, thanks @jsichi for writing the template for this!

jsichi commented:
One last question for @d-michail @jkinable @AlexandruValeanu : should we leave this in connectivity or move it to the new decomposition package created by Alexandru?

AlexandruValeanu commented:
I vote for decomposition. It produces a 'decomposition' so it seems natural.

d-michail commented:
Me too, better move this to the decomposition package.

d-michail commented:
@harmanpa Could you move this class to `org/jgrapht/alg/decomposition`?

harmanpa commented:
@d-michail Done!

jsichi commented:
@d-michail @jkinable any more comments for this one, or can I merge it now?

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/DulmageMendelsohnDecomposition.java:
```diff
@@ -0,0 +1,307 @@
+ * computing both a coarse and a fine Dulmage-Mendelsohn Decomposition.
+ * 
+ * The Dulmage-Mendelsohn Decomposition is based on a maximum-matching of the 
+ * graph G. This implementation uses the Hopcroft-Karp maximum matching 
```
Please wrap all math in $$ signs, see: https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation#math-in-javadoc
$G$

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/DulmageMendelsohnDecomposition.java:
```diff
@@ -0,0 +1,307 @@
+ * A coarse Dulmage-Mendelsohn Decomposition is a partitioning into three 
+ * subsets. Where D is the set of vertices in G that are not matched in the 
+ * maximum matching of G, these subsets are:
+ *   - The vertices in D ∩ U and their neighbors
```
$D \cap U$

harmanpa commented:
@jkinable Thanks, have adjusted math in docs as requested.

My use of this decomposition is similar to in the paper referenced - solution of systems of differential-algebraic-equations (DAEs), i.e. sets of equations containing (time) derivative relationships and algebraic relationships that describe a dynamic system. There is a process, called index-reduction, for converting such a system of equations into a set of ordinary-differential-equations (ODEs) that can be solved with a numerical solver. One part of this process is to determine irreducible blocks, i.e. systems of simultaneous equations, and this is where Dulmage-Mendelsohn is used, based on a bipartite graph of equations and unknowns. If a set dominated by the equations partition is found, then there are overconstrained equations that cannot be solved, likewise if a set dominated by variables is found then it is underconstrained.

harmanpa commented:
@jsichi Good point, I've done that

jsichi commented:
Thanks!
